src/TokenChecks.sol

```diff
@@ -0,0 +1,565 @@
+        uint256 prevWard = TokenLike(_token).wards(address(this));
+        _token.setWard(address(this), 1);
+
+        vm.expectRevert(abi.encodePacked(_contractName, "/invalid-address"));
+        TokenLike(_token).mint(address(0), 1e18);
+        vm.expectRevert(abi.encodePacked(_contractName, "/invalid-address"));
+        TokenLike(_token).mint(_token, 1e18);
+
+        _token.setWard(address(this), prevWard);
+    }
+
+    function checkBurnInsufficientBalance(address _token, string memory _contractName) public {
+        uint256 prevSenderBalance = TokenLike(_token).balanceOf(address(this));
+        deal(_token, address(this), TokenLike(_token).balanceOf(address(this)) + 0.9e18, true);
+
+        vm.expectRevert(abi.encodePacked(_contractName, "/insufficient-balance"));
+        TokenLike(_token).burn(address(this), prevSenderBalance + 1e18);
+    }
+
+    // ************************************************************************************************************
+    // ERC20
+    // ************************************************************************************************************
+
+    function bulkCheckERC20(address _token, string memory _contractName, string memory _tokenName, string memory _symbol, string memory _version, uint8 _decimals) public {
+        checkMetadata(_token, _tokenName, _symbol, _version, _decimals);
+        checkApprove(_token);
+        checkIncreaseAllowance(_token);
+        checkDecreaseAllowance(_token);
+        checkTransfer(_token);
+        checkTransferFrom(_token);
+        checkInfiniteApproveTransferFrom(_token);
+        checkDecreaseAllowanceInsufficientBalance(_token, _contractName);
+        checkTransferBadAddress(_token, _contractName);
+        checkTransferFromBadAddress(_token, _contractName);
+        checkTransferInsufficientBalance(_token, _contractName);
+        checkTransferFromInsufficientAllowance(_token, _contractName);
+        checkTransferFromInsufficientBalance(_token, _contractName);
+    }
+
+    function checkMetadata(address _token, string memory _tokenName, string memory _symbol, string memory _version, uint8 _decimals) public {
+        assertEq(TokenLike(_token).version(), _version); // Note that this is not part of the ERC20 standard
+        assertEq(TokenLike(_token).name(), _tokenName);
+        assertEq(TokenLike(_token).symbol(), _symbol);
+        assertEq(TokenLike(_token).decimals(), _decimals);
+    }
+
+    function checkApprove(address _token) public {
+        vm.expectEmit(true, true, true, true);
+        emit Approval(address(this), address(0xBEEF), 1e18);
+
+        assertTrue(TokenLike(_token).approve(address(0xBEEF), 1e18));
+
+        assertEq(TokenLike(_token).allowance(address(this), address(0xBEEF)), 1e18);
+    }
+
+    function checkIncreaseAllowance(address _token) public {
+        uint256 prevAllowance = TokenLike(_token).allowance(address(this), address(0xBEEF));
```

Comment:
IMO we should force some prev allowance so we make sure increase allowance acts as that and not like regular approve.

Reply:
Done